core/src/main/scala/com/gu/tableversions/core/InMemoryTableVersions.scala

```diff
@@ -15,9 +17,22 @@ import com.gu.tableversions.core.util.RichRef._
 class InMemoryTableVersions[F[_]] private (allUpdates: Ref[F, TableUpdates])(implicit F: Sync[F])
```

Comment:
Disclaimer: sorry, I know it's not something that's been introduced in this PR, but...

We've got quite a bit of business logic in this concrete implementation that looks likely to me to be the only sensible implementation of this API. I think we can hoist it up to `TableVersions` and achieve a few benefits for doing so:

```scala
/**
  * This defines the interface for querying and updating table version information tracked by the system.
  */
trait TableVersions[F[_]] {

  /**
    * Start tracking version information for given table.
    * This must be called before any other operations can be performed on this table.
    */
  def init(table: TableName, isSnapshot: Boolean, userId: UserId, message: UpdateMessage, timestamp: Instant): F[Unit]

  /** Get a description of which version to write to next for the given partitions of a table. */
  def nextVersions(table: TableName, partitions: List[Partition]): F[Map[Partition, Version]]

  /**
    * Update partition versions to the given versions.
    * This performs no checking if data has been written to the associated paths but it will verify that these versions
    * 1) haven't been committed before and 2) these are the next versions to be committed for each of the partitions.
    */
  def commit(table: TableName, update: TableVersions.TableUpdate): F[CommitResult]

  def updates: fs2.Stream[F, TableUpdate]

  // Implementors are free to provide a more efficient override
  def updates(forTable: TableName): fs2.Stream[F, TableUpdate] =
    updates.filter(_.table == forTable)

  /** Get details about partition versions in a table. */
  def currentVersion(table: TableName): F[TableVersion] =
    updates(table).fold(doMyThingWithTableOperations)
}
```

Now implementors are only required to provide the primitives (`updates` and potentially `updates(forTable)`), and you're able to give them a useful bit of the API for free (`currentVersion`).

Plus it doesn't seem unreasonable to me to expose the updates; I'm sure it's useful. And by using an fs2 `Stream`, there's no requirement now to pass potentially very large lists of updates around strictly.

See what you think - I'm just thinking out loud here and it certainly wouldn't need to go in with this PR regardless!

Reply:
I like the idea of using fs2 `Stream`s instead of `List`s for these. I'll add that to the list.

W.r.t. pulling out potential common code, I like the suggestion, though I'd probably wait until we did add another implementation.

Regarding your specific suggested code - I'm not sure it makes sense to make implementors implement a stream of all update events across tables? This wouldn't really be used for anything as it stands (other than the non-efficient default implementation of `updates(forTable: TableName)`, and it may be painful to do for some backends. For example if we have an implementation using DynamoDB with a separate document per table as its schema, it would have to do a query per table for to provide this.

Reply:
> W.r.t. pulling out potential common code, I like the suggestion, though I'd probably wait until we did add another implementation.

That's cool, no rush!

FTR, that's not really my motivation here though (the pulling out of potential common code). 

I think it would make `TableVersions` more useful and cohesive as an abstraction: the operations defined on it would have stronger relationships with one another and it would provide better guarantees.

> I'm not sure it makes sense to make implementors implement a stream of all update events across tables?

That's fair. Maybe just `updates(forTable: TableName): Stream[F, TableUpdate]` then.